README.md
# Development environment 3 (devenv3)

Окружение разработки 3 предназначено для удобной локальной разработки приложений.

Для работы Окружения необходимы следующие зависимости:

- bash
- docker
- docker-compose не ниже **1.12.0** версии

_bash_ и _docker_ устанавливаются глобально через пакетный менеджер Вашего дистрибутива,
_docker-compose_ можно установить также глобально через пакетный менеджер,
либо локально через **pip** (Python Package manager).

Пример установки зависимостей в Debian/Ubuntu-производных дистрибутивах:

```sh
$ sudo apt-get install bash docker docker-compose
```

Далее необходимо склонировать Окружение разработки в любую папку, например в домашнюю:

```sh
$ git clone https://github.com/abra7134/devenv3 ~/devenv3
```

## Основные команды

Для управления Окружением разработки подготовлен специальный **BASH**-скрипт `devenv3.bash`.
Скрипт поддерживает следующие команды:

- _build_ - сборка необходимых Docker-образов
- _down_ - удаление контейнеров Окружения Разработки (необходимо при обновлениях)
- _help_ - помощник по использованию скрипта
- _ls_ - вывод списка всех приложений с указанием используемой версии **PHP**-интерпретатора,
выбранного индексного файла, а также выбранной **GIT**-ветки или **Mercurial**-ветки
- _init_ - первоначальная инициализация Окружения Разработки (можно запускать несколько раз)
- _rm_ - удаление приложений или их алиасов
- _run_ - запуск команды внутри Окружения Разработки (например: `composer`, `php yii`, `php artisan` и т.д.)
- _run_at_ - запуск команды в указанном приложении внутри Окружения Разработки (удобно для документации)
- _set_at_ - управление различными параметрами работы приложения
(установка версии **PHP**, использование XDebug расширения, добавление алиаса)
- _up_ - локальный запуск Окружения Разработки с предварительным запуском _down_ команды

#### Инициализация окружения разработки

Перед началом работы необходимо проинициалировать окружение, запустив скрипт:

```sh
$ cd ~/devenv3  # указывается путь куда было склонировано Окружение разработки
$ bash devenv3.bash init
```

Инициализация выполнит:

- конфигурацию файла настроек `.env`;
- создаст `~/www/` папку при необходимости;
- добавит алиасы скрипта в `.bashrc` и `.zshrc` файлы.

После этого необходимо запустить новую консоль и скрипт будет доступен по следующими коротким именам:
```sh
$ de3
$ denv3
$ devenv3
```

#### Сборка необходимых `Docker`-образов

Далее необходимо собрать локальные `Docker`-образы. Операция может занять некоторое время.
```sh
$ de3 build
```

#### Локальный запуск Окружения разработки

И, наконец, остаётся запустить Окружение разработки. Команду рекомендуется запускать в отдельном терминале,
для удобного наблюдения за логами:
```sh
$ de3 up
```

## Схема работы с PHP-приложениями

#### Общая информация

Для разработки и тестирования веб-приложений в Окружении разработки доступны веб-сервер **nginx** и обработчики
**php5.6-fpm**, **php7.1-fpm**, **php7.2-fpm** и **php7.4-fpm** с необходимыми преднастройками
(почти идентично повторяющими настройки боевых серверов). Окружение разработки доступно под локальным доменом:

    *.localhost

Для хранения веб-приложений на этапе инициализации Окружения разработки была создана отдельная
папка `~/www/` в домашней директории.

#### Структура `~/www/` папки
```
~/www/[default/]
~/www/[catchall/]
~/www/имя_приложения-1/
~/www/имя_приложения-2/
~/www/...
~/www/имя_приложения-X/
```
Папка `www/` в домашней директории предназначена для хранения веб-приложений.
Имя любой папки здесь сразу отображается в имя хоста.
Т.е. для организации доступа к новому веб-приложению достаточно просто **создать папку** с нужным названием
и **скопировать** туда приложение!

Приложение будет сразу доступно по локальному адресу: `http://имя_приложения.localhost`
(например: `http://api_client.localhost`).

`имя_приложения` может содержать буквы, цифры, тире и знак подчеркивания (в regexp нотации `[[:alnum:]_-]`).

Также существуют два специальных имени: `default` и `catchall` с нестандартным поведением:

* Приложение в папке `default` будет обрабатывать запросы, поступившие на короткий адрес
(без указания имени приложения), т.е. просто на `http://localhost`;
* А приложение в папке `catchall` будет обрабатывать запросы всех **не существующих** приложений.

Поиск индексных файлов `index.htm`, `index.html`, `index.php` осуществляется в следующих папках
(что найдет позднее - то и обработает):
```
~/www/имя_приложения/
~/www/имя_приложения/web/
~/www/имя_приложения/api/web/
~/www/имя_приложения/public/
```
Для добавления новых типов индексных файлов, а также папок для их поиска следует обратиться к администратору.

PS: Изменение прав доступа к файлам не требуется, т.к. все файлы хранятся и обрабатываются под личной учетной записью
(!!! никаких `chmod 777` использовать не нужно - это признак дурного тона).

#### Алиасы

Окружение разработки поддерживает *алисы*, т.е. альтернативные названия приложений.
Для этого достаточно создать символическую ссылку. Сделать это можно с помощью команды `de3 set_at`:

Например, мы хотим, чтобы приложение `http://api_client.localhost` было также доступно и как `http://ca.localhost`:
```sh
$ de3 set_at api_client alias ca
```

Для правильного функционирования сайта на основе `yii-advanced` темплейта рекомендуется создать следующие алиасы
(предположим он располагается в папке `site`):
```sh
$ de3 set_at site/backend  alias admin
$ de3 set_at site/frontend alias catchall
$ de3 set_at site/landing  alias landing
$ de3 set_at site/landing  alias default
```

#### PHP

В Окружении разработки доступны **5.6**, **7.1**, **7.2** и **7.4** версии **PHP**-интерпретатора,
однако по умолчанию будет использоваться **5.6** версия.
Для указания необходимой версии для конкретного приложения достаточно воспользоваться той же командой `de3 set_at`:
```sh
$ de3 set_at имя_приложения php 5.6
$ de3 set_at имя_приложения php 7.1
```

Команда запишет специальную метку в папку приложения
(а именно пустые файлы `.profile_php7.1`, `.profile_php7.2` или `.profile_php7.4` (обратите внимание на точку в начале).

#### XDebug

Для включения/отключения удаленной трассировки с помощью **PHP**-расширения **xdebug** можно также воспользоваться
командой `de3 set_at`, которая запишет или удалит специальный пустой файл `.profile_xdebug` в папке приложения.

Наличие этого файла даст указание веб-серверу использовать **PHP**-интерпретатор с параметрами:
`xdebug.remote_enable=1`, `xdebug.remote_connect_back=1`, которые разрешают удаленную трассировку,
а также увеличенным максимальным временем исполнения **PHP**-скрипта *до 120 секунд* (в обычном режиме *только 10 секунд*).

Пример использования:
```sh
$ de3 set_at имя_приложения xdebug on
$ de3 set_at имя_приложения xdebug off
```

Если необходимо вместо **remote_connect_back** режима использовать **remote_autostart**,
то можно воспользоваться переопределением в `.user.ini` файле,
который необходимо создать в папке с индексным файлом
(конкретное расположение индексного файла можно узнать запустив команду Окружения разработки `ls`)
со следующим содержимым:
```ini
xdebug.remote_connect_back=0
xdebug.remote_autostart=1
xdebug.remote_host=адрес_хоста_с_xdebug_отладчиком
xdebug.remote_port=порт_xdebug_отладчика
```

#### HTTP-заголовки

В большинстве случаев на боевых серверах установлены пограничные балансировщики (в том числе и **SSL**-терминаторы),
которые запрос из Интернета траслируют уже во внутреннюю сеть приложения (кластера). Тем самым **PHP**-переменная
`REMOTE_ADDR` будет содержать в себе не **IP**-адрес реального клиента, который послал запрос, а **IP**-адрес
конкретного балансировщика. Оригинальный же **IP**-адрес будет записан балансировщиком в **HTTP**-заголовки
`X-Forwarded-For` и `X-Real-Ip`. А в заголовки `X-Forwarded-Proto` и `X-Real-Ip` запишутся, соответственно,
изначальный протокол и оригинальный TCP-порт клиента.

В Окружении разработки балансировщик отсутствует.
Но для удобства разработки эти заголовки устанавливаются в принудительном порядке **HTTP**-сервером.

`X-Real-Ip` будет содержать в себе IP-адрес клиента (в окружении разработки будет совпадать с **PHP**-переменной `REMOTE_ADDR`)
`X-Real-Port` будет содержать TCP-порт клиента, соответственно.
`X-Forwarded-For` и `X-Forwarded-Proto` - цепочку IP-адресов и изначальный протокол соответственно.

Например:
```
X-Real-Ip: 192.168.1.23
X-Real-Port: 56723
X-Forwarded-For: 192.168.0.10, 192.168.1.23
X-Forwarded-Proto: http
```

#### Вывод списка всех PHP-приложений

С помощью команды `de3 ls` можно посмотреть список всех установленных **PHP**-приложений,
а также получить дополнительную информацию о версии **PHP**-интерпретатора, выбранном индексном файле,
а также наименовании текущей **GIT**-ветки или **Mercurial**-ветки.

Колонка `TP` указывает **тип** приложения:
`->` означает алиас, `==` означает прямое отображение имени приложения в имя папки.

```sh
$ de3 ls
...
NAME          URL                          TP HOME               INDEX FILE           PHP         BRANCH
api-1         http://api-1.localhost/      == api-1/             index.php            7.4         -

CHANGELOG.md
# Development environment 3 (devenv3)

## История изменений

### v0.4.4 (2022/07/15)
- **docker-compose.yml**: Исправлен проброс build-переменных при использовании YAML-overrides

### v0.4.3 (2022/07/06)
- Большинство контейнеров теперь используют Debian **10** в качестве базового слоя

### v0.4.2 (2022/06/28)

- **devenv3.bash**: В команду `build` добавлена проверка существования `.env` файла
- **php-fpm-72**, **php-fpm-74**: Восстановлена работа трассировки (через `xdebug`-расширение)
- **php-fpm-**: Добавлен пакет `imagemagick` для работы с изображениями

### v0.4.1 (2022/06/13)

- Добавлена поддержка `php-fpm-74` обработчика
- **composer**: Composer обновлен до версии **1.10.26**

### v0.4 (2021/12/16)

###### Значительные изменения:
- Все контейнеры теперь используют Debian **9.13** в качестве базового слоя
- **devenv3.bash**: Команда `init` теперь добавляет `de3`-алиасы и для **zsh**
- **devenv3.bash**: В команду `ls` добавлена поддержка отображения **git**-веток

###### Остальные изменения:
- **nginx**: Добавлена принудительная установка `HTTP`-заголовков `X-Forwarded-Proto`, `X-Real-Ip` и `X-Real-Port`
- **devenv3.bash**: Исправлена работа команды `init` при отсутствующем `~/.bashrc` файле,
  добавлен вывод ошибки при невозможности записи `~/.bashrc` и `.env` файлов
- **devenv3.bash**: заголовок **shebang** заменен на более совместимый вариант
- **php-fpm-**: Добавлен пакет `mercurial` для скачивания зависимостей через `composer`
- **devenv3.bash**: Аргументы команды `build` теперь пробрасываются до `docker-compose`
- **devenv3.bash**: Стандартизирован список допустимых символов в названии приложенения `[[:alnum:]\_-]`
- **devenv3.bash**: Улучшена логика обработки ситуаций, когда домашняя папка приложения находится
  за пределами папки с приложениями
- **devenv3.bash**: Теперь показывается ошибка, если указана несуществующая команда (ранее показывался `help`)

### v0.3 (2019/04/18)

###### Значительные изменения:
- **devenv3.bash**: Добавлена новая команда `rm` для удаления приложений или их алиасов
- **devenv3.bash**: Добавлена новая команда `set_at` для управления параметрами приложений:
  можно указать нужную версию PHP-интерпретатора, использовать или нет XDebug-расширение, а также добавить алиас
- **docker-compose.yml**: В качестве временного решения резолвинга адресов вида `http://*.localhost`
  внутри Окружения разработки (например для прямого обращения `PHP`-приложений к `PHP`-приложениям)
  добавлен контейнер с DNS-сервером `dnsmasq` с полным логированием всех DNS-запросов

###### Остальные изменения:
- **devenv3.bash**: В команду `ls` добавлена новая колонка `TP` (TYPE) с отображением типа приложения
- **devenv3.bash**: В команде `ls` исправлено отображение пути до индексного файла (убран лишний '/')
  в случае его нахождения в корне папки приложения
- **devenv3.bash**: В команду `up` добавлен предварительный запуск `down` команды, что удобнее при обновлении
- **nginx**: Добавлена принудительная установка `HTTP`-заголовка `X-Forwarded-For` для `PHP`-приложений
- **php-fpm-**: Восстановлена работоспособность `PHP`-трассировки (для медленных запросов и исключений)
- **php-fpm-56**, **php-fpm-71**: Добавлено недостающее расширение `php-zip`
- **README.md**: Убрана путаница с использованием синонимов **проект** и **приложение**,
  а также другие мелкие улучшения читабельности и понимания
- Основная разработка теперь ведется в отдельной ветке `develop`

### v0.2 (2019/03/28)

###### Значительные изменения:
- `Makefile` заменен на BASH-скрипт `devenv3.bash` с улучшенной архитектурой и новыми командами:
- - down
- - ls
- - run
- - run_at

###### Остальные изменения:
- В контейнеры добавлен проброс `/dev/log` для корректной работы `syslog`
- Удалено использование неработающего репозитария 'jessie-updates' (устарел) из всех контейнеров
- Добавлен `UPGRADING.md`
- В отдельный контейнер добавлен `composer` последней версии
- Директории `configs` с конфигурацией для `Docker`-образов переименованы в `root`
- **php-fpm-***: Добавлены пакеты `git`, `libdbi-perl`, `libdbd-mysql-perl`, `mysql-client-5.5`, `unzip` для работы миграций
- **nginx**: Добавлен аргумент `NGINX_VERSION` в `Dockerfile`
- Добавлен `CHANGELOG.md`

### v0.1 (2019/02/05)

- Добавлен `README.md`
- Уменьшен размер `opcache` для `PHP`-обработчиков с 512 до 256 Мбайт
- Добавлена поддержка `.profile-xdebug` и `php-fpm-71`, `php-fpm-72` обработчиков
- Папка приложений перемещена из `/opt/www` в `~/www`
- Первая внутренняя версия


devenv3.bash
#!/usr/bin/env bash

DEVENV3_VERSION="0.4.4"
DEVENV3_MAINTAINER_EMAIL="lekomtsev@unix-mastery.ru"

DEVENV3_APP_DIR="${HOME}/www"
DEVENV3_APP_DOMAIN="localhost"
DEVENV3_HOME_DIR=$(dirname $(realpath "${0}"))
DEVENV3_FILENAME=$(basename "${0}")
DEVENV3_ENV_FILEPATH="${DEVENV3_HOME_DIR}/.env"

DEVENV3_ALIASES=("de3" "denv3" "devenv3")
DEVENV3_ALIAS="${_devenv3_alias:-${DEVENV3_FILENAME}}"

BASHRC_PATH="${HOME}/.bashrc"
ZSHRC_PATH="${HOME}/.zshrc"

# Disable file name generation (globbing)
set -f

function _print {
  local print_type="${1}"
  shift

  if [[ -z "${1}" ]]; then
    return 0
  fi

  case "${print_type}" in
    "info" )
      local info_message
      for info_message in "${@}"; do
        printf "!!! ${info_message}\n"
      done
      echo
      ;;
    "progress" )
      local progress_message
      for progress_message in "${@}"; do
        printf "[*] ${progress_message} ...\n"
      done
      ;;
    * )
      {
        local ident="!!! ${print_type^^}:"
        local ident_length="${#ident}"
        echo
        local error_message
        for error_message in "${@}"; do
          printf "%-${ident_length}s %s\n" "${ident}" "${error_message}"
          ident=""
        done
        echo
      } >&2
      ;;
  esac
}

function error {
  _print error "${@}"
  exit 1
}

function info {
  _print info "${@}"
}

function internal {
  _print internal \
    "Problem in '${FUNCNAME[2]}' -> '${FUNCNAME[1]}' function call at ${BASH_LINENO[1]} line:" \
    "${@}"
  exit 2
}

function progress {
  _print progress "${@}"
}

function warning {
  _print warning "${@}"
  exit 0
}

function run {
  local command="${1}"

  if [[ -z "${command}" ]]; then
    internal "Function parameters needs to be specified"
  fi

  local command_type="$(type -t "${command}")"
  if [[ "${command_type}" == "file" ]]; then
    shift
    "${command}" "${@}" \
      || error "Failed of '${command}' command running" \
               "Please check and try again, or address the administrator"
  elif [[ -z "${command_type}" ]]; then
    error \
      "The command '${command}' must be exists at your system" \
      "Please install it from the package manager of your OS"
  else
    internal "The specified command '${command}' needs to be 'file' type, not '${command_type}'"
  fi
}

function run_inside_de3 {
  pushd "${DEVENV3_HOME_DIR}" >/dev/null
  run "${@}"
  popd >/dev/null
}

function main_footer {
  cat <<END_INFO

Development Environment 3 v${DEVENV3_VERSION}
Any questions please send to: ${DEVENV3_MAINTAINER_EMAIL}

DevEnv3 home directory: ${DEVENV3_HOME_DIR}
Applications directory: ${DEVENV3_APP_DIR}

END_INFO
}

function command_build {
  if [[ "${1}" == "description" ]]; then
    echo "Build Docker images for the DevEnv3 running"
    return 0
  fi

  if [ ! -s "${DEVENV3_ENV_FILEPATH}" ]; then
    warning \
      "Correct images building requires an '.env' file at DevEnv3 home directory" \
      "Please run '${DEVENV3_ALIAS} build' command at first for correct writing this file"
  fi

  info \
    "This command run a building of Docker images used in this environment" \
    "It can take some time"

  progress "Starting 'docker-compose build' command"
  run_inside_de3 \
    docker-compose build "${@}"
  progress "Done"
}

function command_down {
  if [[ "${1}" == "description" ]]; then
    echo "Destroy containers and other Docker things of the DevEnv3 (situable for a upgrading)"
    return 0
  fi

  progress "Starting 'docker-compose down' command"
  run_inside_de3 \
    docker-compose down
  progress "Done"
}

function command_help {
  if [[ "${1}" == "description" ]]; then
    echo "Print this help"
    return 0
  fi

  echo "Usage:"

  local devenv3_alias
  for devenv3_alias in ${DEVENV3_ALIASES[@]}; do
    echo "  ${devenv3_alias} COMMAND [OPTIONS]"
  done
  echo
  echo "Commands:"
  local function_name
  for function_name in $(compgen -A function); do
    if [[ "${function_name}" =~ ^command_ ]]; then
      printf "  %-10s %s\n" "${function_name#command_}" "$(${function_name} description)"
    fi
  done
  echo
  echo "Run '${DEVENV3_ALIAS} COMMAND --help' for more information on a command (IN A FUTURE RELEASE)."
  echo
  exit 0
}

function command_init {
  if [[ "${1}" == "description" ]]; then
    echo "Append DevEnv3 aliases to .bashrc, .zshrc and other initializations"
    return 0
  fi

  local begin_string="### DevEnv3 aliases BEGIN ###"
  local end_string="### DevEnv3 aliases END ###"
  local rc_path

  progress "Creating Applications directory"
  run mkdir --parents \
    "${DEVENV3_APP_DIR}"

  progress "Writing '${DEVENV3_ENV_FILEPATH}' file"
  {
    echo "USER_ID=`id --user`"
    echo "GROUP_ID=`id --group`"
  } > "${DEVENV3_ENV_FILEPATH}" \
    || error "Failed to write file"

  for rc_path in "${BASHRC_PATH}" "${ZSHRC_PATH}"; do
    if [ -f "${rc_path}" ]; then
      progress "Appending DevEnv3 aliases to '${rc_path}'"
      run sed --in-place \
        "/${begin_string}/,/${end_string}/d" \
        "${rc_path}"
    elif [ ! -h "${rc_path}" -a ! -e "${rc_path}" ]; then
      progress "Writing DevEnv3 aliases to '${rc_path}'"
    else
      error "Failed to write '${rc_path}' because it's not a regular file"
    fi
    {
      echo "${begin_string}"
      local devenv3_alias
      for devenv3_alias in ${DEVENV3_ALIASES[@]}; do
        echo "alias ${devenv3_alias}=\"_devenv3_alias=${devenv3_alias} /usr/bin/env bash \\\"${DEVENV3_HOME_DIR}/${DEVENV3_FILENAME}\\\"\""
      done
      echo "${end_string}"
    } >> "${rc_path}" \
      || error "Failed to write file"
  done

  progress "Done"

  warning "To work of DevEnv3 aliases properly the another copy of terminal to be run" \
          "OR run this command in BASH: source ${BASHRC_PATH}" \
          "                    in  ZSH: source ${ZSHRC_PATH}"
}

function command_ls {
  if [[ "${1}" == "description" ]]; then
    echo "List all installed applications"
    return 0
  fi

  local print_format="%-20s %-40s %-2s %-20s %-20s %-12s %-10s\n"
  printf "${print_format}" \
    "APP NAME" "URL" "TP" "HOME" "INDEX FILE" "PHP" "BRANCH"

  local app_{branch,branch_temp,dir,home,index_file,name,php_version,type,url}
  local index_{dir,file}

  set +f
  for app_dir in "${DEVENV3_APP_DIR}/"*; do
    if [[    ! -d "${app_dir}" \
          && ! -h "${app_dir}" ]]; then
      continue
    fi

    app_name="${app_dir##*/}"

    if [[ ! "${app_name}" =~ ^[[:alnum:]_-]+$ ]]; then
      app_url="(WRONG NAME)"
    else
      case "${app_name}" in
        "catchall" )
          app_url="http://*.${DEVENV3_APP_DOMAIN}/"
          ;;
       "default" )
          app_url="http://${DEVENV3_APP_DOMAIN}/"
          ;;
        * )
          app_url="http://${app_name}.${DEVENV3_APP_DOMAIN}/"
          ;;
      esac
    fi

    app_home=$(
      run realpath \
        --relative-base="${DEVENV3_APP_DIR}" \
        "${app_dir}"
    )
    app_branch="-"
    app_index_file="-"
    app_type="->"
    app_php_version="-"

    # Don't process application if realdir begin with / that is OUTSIDE applications directory
    if [[ "${app_home::1}" == "/" ]]; then
      app_home="(OUTSIDE)"
    # And also if resolved directory is not exists
    elif [[ ! -d "${DEVENV3_APP_DIR}/${app_home}" ]]; then
      app_home="(MISSING)"
    else
      if [[ "${app_name}" == "${app_home}" ]]; then
        app_type="=="
      fi

      # Overwrite a app_dir with real directory
      app_dir="${DEVENV3_APP_DIR}/${app_home}"
      # For a pretty printing
      app_home+="/"

      for app_php_version in "7.4" "7.2" "7.1" "5.6"; do
        if [[ -f "${app_dir}/.profile_php${app_php_version}" ]]; then
          break
        fi
      done
      if [[ -f "${app_dir}/.profile_xdebug" ]]; then
        app_php_version+="+xdebug"
      fi

      for index_dir in "public/" "api/web/" "web/" ""; do
        if [[ -d "${app_dir}/${index_dir}" ]]; then
          break
        fi
      done
      for index_file in "index.htm" "index.html" "index.php" ""; do
        if [[ -f "${app_dir}/${index_dir}${index_file}" ]]; then
          break
        fi
      done
      if [[ -z "${index_file}" ]]; then
        app_index_file="(NOT FOUND)"
      else
        app_index_file="${index_dir}${index_file}"
      fi

      index_dir="${app_dir}"
      # Walk to all upper directories when a searching of branch name like 'hg' command :)
      while [[ "${index_dir}" != "${DEVENV3_APP_DIR}" ]]; do
        if [[ -s "${index_dir}/.git/HEAD" ]]; then
          read app_branch_temp <"${index_dir}/.git/HEAD"
          # FIXME: realise the better branch name checker
          if [[ "${app_branch_temp}" =~ ^ref:\ refs/heads/([[:alnum:]\._-]+)$ ]]; then
             app_branch="git:${BASH_REMATCH[1]}"
             break
          fi
        elif [[ -s "${index_dir}/.hg/branch" ]]; then
          read app_branch_temp <"${index_dir}/.hg/branch"
          # FIXME: realise the better branch name checker
          if [[ "${app_branch_temp}" =~ ^[[:alnum:]\._-]+$ ]]; then
            app_branch="hg:${app_branch_temp}"
            break
          fi
        fi
        index_dir="${index_dir%/*}"
      done
    fi

    printf "${print_format}" \
      "${app_name}" \
      "${app_url}" \
      "${app_type}" \
      "${app_home}" \
      "${app_index_file}" \
      "${app_php_version}" \
      "${app_branch}"
  done

  echo
  info "Legend:" \
       "  TP - type of application: '->' is alias, '==' is a direct transformation of an application name to a folder name"
}


function command_rm {
  if [[ "${1}" == "description" ]]; then
    echo "Remove unnecessary applications or their aliases"
    return 0
  fi

  local command="${1}"
  if [[ -z "${command}" || "${command}" == "--help" ]]; then
    warning \
      "Please specify applications which must be removed" \
      "Usage: ${DEVENV3_ALIAS} ${command_name} <application_name1> [application_name2]..."
  fi

  local app_{dir,home,name}
  for app_name in "${@}"; do
    app_dir="${DEVENV3_APP_DIR}/${app_name}"

    if [ -h "${app_dir}" ]; then
      app_home=$(
        run realpath \
          --relative-base="${DEVENV3_APP_DIR}" \
          "${app_dir}"
      )

      # Remove an alias only at inside of DEVENV3_APP_DIR directory
      if [[ "${app_home::1}" != "/" ]]; then
        progress "Remove an alias '${app_name}' of '${app_home}' application"
        run rm "${app_dir}"
      else
        progress "Skipping a removing an alias '${app_name}' because it linked to outside of Applications directory"
      fi

    elif [ -d "${app_dir}" ]; then
      progress "Remove an application '${app_name}'"
      run rm --recursive "${app_dir}"
    fi
  done

  progress "Done"
}

function command_run {
  if [[ "${1}" == "description" ]]; then
    echo "Run any command inside the DevEnv3 (for example: composer, php and etc)"
    return 0
  fi

  local command="${1}"
  if [[ -z "${command}" || "${command}" == "--help" ]]; then
    warning \
      "Please specify a running command" \
      "Usage: ${DEVENV3_ALIAS} ${command_name} <command> [parameters]"
  fi

  shift
  local pwd_rel_dir="${PWD#${DEVENV3_APP_DIR}/}"
  if [[ "${PWD}" == "${pwd_rel_dir}" ]]; then
    error "The '${DEVENV3_ALIAS} ${command_name}' command must be runned inside any application directory!"
  fi

  command_run_at "${pwd_rel_dir}" "${command}" "${@}"
}

function command_run_at {
  if [[ "${1}" == "description" ]]; then
    echo "Run any command at selected application inside the DevEnv3 (for example: composer, php and etc)"
    return 0
  fi

  local app_path="${1}"
  if [[ -z "${app_path}" || "${app_path}" == "--help" ]]; then
    warning \
      "Please specify an application name where the command will run" \
      "Usage: ${DEVENV3_ALIAS} ${command_name} <application_name> <command> [parameters]"
  fi

  shift
  local app_name="${app_path%%/*}"
  local app_dir="${DEVENV3_APP_DIR}/${app_name}"
  if [[ ! -d "${app_dir}" ]]; then
    error \
      "The specified '${app_name}' application is not exists, please check and try again" \
      "To view list of all installed applications please use the command:" \
      "$ de3 ls"
  fi

  local command="${1}"
  if [[ -z "${command}" || "${command}" == "--help" ]]; then
    warning \
      "Please specify a running command" \
      "Usage: ${DEVENV3_ALIAS} ${command_name} <application_name> <command> [parameters]"
  fi
  shift

  local app_php_version
  for app_php_version in "7.4" "7.2" "7.1" "5.6"; do
    if [[ -f "${app_dir}/.profile_php${app_php_version}" ]]; then
      break
    fi
  done

  local container_name="php-fpm-${app_php_version/./}"  # With remove a . (dot) from php version
  progress "Checking of existing the '${container_name}' container"
  # The docker-compose v1.12.0 '-q' flag have only
  local container_id="$(run_inside_de3 docker-compose ps -q "${container_name}")"
  if [[    -z "${container_id}" \
        || -z "$(run docker ps --quiet \
                               --filter "status=running" \
                               --filter "id=${container_id}")" ]]; then
    error \
      "The necessary container have not 'running' state" \
      "Please run the DevEnv3 by '${DEVENV3_ALIAS} up' command"
  fi

  progress "Run a '${command}' command at application '${app_name}' using '${container_name}' container"
  # Use exec direct from docker to avoid a problem with BASH "${@}" expansion
  # Because with docker-compose command it is necessary to use next command:
  # /bin/sh -c "cd /www/${pwd_rel_dir}; ${command} ${@}"
  # 1. Hack with WORKDIR is using because Setting workdir for exec is not supported in API < 1.35 (1.22)
  # 2. This form of ${@} expanded only one parameter :(
  run_inside_de3 \
    docker exec \
      --interactive \
      --tty \
      --user www-data \
      --workdir "/www/${app_path}" \
      "${container_id}" \
      "${command}" "${@}"
}

function command_set_at {
  if [[ "${1}" == "description" ]]; then
    echo "Set any parameters at selected application (for example: PHP version, enable/disable Xdebug, ...)"
    return 0
  fi

  local app_name="${1}"
  if [[ -z "${app_name}" || "${app_name}" == "--help" ]]; then
    warning \
      "Please specify an application name where set parameters" \
      "Usage: ${DEVENV3_ALIAS} ${command_name} <application_name> <parameter_name> <parameter_value>"
  fi

  local app_path="${DEVENV3_APP_DIR}/${app_name}"
  if [[ ! -d "${app_path}" && ! -h "${app_path}" ]]; then
    error \
      "The specified '${app_name}' application is not exists, please check and try again" \
      "To view list of all installed applications please use the command:" \
      "$ de3 ls"
  fi

  # ${app_home} is relative to ${DEVENV3_APP_DIR} directory
  local app_home=$(
    run \
      realpath \
        --relative-base="${DEVENV3_APP_DIR}" \
        "${app_path}"
  )
  if [[ "${app_home::1}" == "/" ]]; then
    error \
      "The specified '${app_name}' application has a home which is placed outside Applications directory" \
      "Please specify an another application name!" \
      "To view list of all installed applications please use the command:" \
      "$ de3 ls"
  fi
  local app_real_path="${DEVENV3_APP_DIR}/${app_home}"

  shift
  local parameter_name="${1}"
  local parameter_value="${2}"
  case "${parameter_name}" in
    "alias" | "Alias" | "ALIAS" )
      if [[ ! "${parameter_value}" =~ ^[[:alnum:]_-]+$ ]]; then
        error "Please specify a correct alias name with symbols 'a-z', 'A-Z', '_', '-' and digits '0-9' only supported"
      fi

      local app_alias_path="${DEVENV3_APP_DIR}/${parameter_value}"
      if [[ -d "${app_alias_path}" ]]; then
        error "The application '${parameter_value}' is already exists, an alias creation is impossible, skipping"
      elif [[ -h "${app_alias_path}" ]]; then
        error "The alias '${parameter_value}' is already exists, please specify an another name"
      fi

      if [[ "${app_name}" != "${app_home}" ]]; then
        progress "Create a symlink '${parameter_value}' -> '${app_path}' (which is alias to '${app_home}')"
      else
        progress "Create a symlink '${parameter_value}' -> '${app_path}'"
      fi

      run ln --symbolic \
        "${app_path}" \
        "${app_alias_path}"

      progress "Done"
      ;;
    "php" | "Php" | "PHP" )
      if [[ ! "${parameter_value}" =~ ^(5\.6|7\.1|7\.2|7\.4)$ ]]; then
        error "Please specify a correct version of PHP-interpreter: '5.6', '7.1', '7.2' or '7.4' only supported"
      fi

      local app_php_version
      local app_php_profile_file
      for app_php_version in "7.4" "7.2" "7.1" "5.6"; do
        app_php_profile_file="${app_real_path}/.profile_php${app_php_version}"
        if [[ -f "${app_php_profile_file}" ]]; then
          break
        fi
      done

      if [[ "${parameter_value}" == "${app_php_version}" ]]; then
        warning "The application '${app_name}' has already use PHP ${app_php_version} version, skipping..."
      fi

      if [[ "${app_php_version}" != "5.6" ]]; then
        progress "Remove the old '${app_php_profile_file}' file"
        run rm --force \
          "${app_php_profile_file}"
      fi

      if [[ "${parameter_value}" != "5.6" ]]; then
        app_php_profile_file="${app_real_path}/.profile_php${parameter_value}"
        progress "Write the new '${app_php_profile_file}' empty file"
        run touch \
          "${app_php_profile_file}"
      fi

      progress "Done"
      ;;
    "xdebug" | "Xdebug" | "XDEBUG" )
      if [[ ! "${parameter_value}" =~ ^(on|On|ON|off|Off|OFF)$ ]]; then
        error "Please specify a correct value of parameter: 'on' and 'off' only supported"
      fi

      local app_php_xdebug="off"
      local app_php_xdebug_file="${app_real_path}/.profile_xdebug"
      if [[ -f "${app_php_xdebug_file}" ]]; then
        app_php_xdebug="on"
      fi

      # Compared with a lowest case version of ${parameter_value}
      case "${app_php_xdebug} -> ${parameter_value,,}" in
        "on -> on" )
          warning "The application '${app_name}' already use XDebug extension, skipping..."
          ;;
        "off -> off" )
          warning "The application '${app_name}' already don't use XDebug extension, skipping..."
          ;;
        "on -> off" )
          progress "Remove the old '${app_php_xdebug_file}' file"
          run rm --force \
            "${app_php_xdebug_file}"
          ;;
        "off -> on" )
          progress "Write the new '${app_php_xdebug_file}' empty file"
          run touch \
            "${app_php_xdebug_file}"
          ;;
      esac

      progress "Done"
      ;;
    * )
      warning \
        "Please specify a parameter and its value which will be set" \
        "Usage: ${DEVENV3_ALIAS} ${command_name} <application_name> <parameter_name> <parameter_value>" \
        "" \
        "Parameters:  alias <alias_name>      Set an alias (alternavite name)" \
        "             php (5.6|7.1|7.2|7.4)   Set a version of PHP-interpreter which will be used" \
        "             xdebug (on|off)         Enable or disable usage XDebug extension with PHP"
      ;;
  esac

}

function command_up {
  if [[ "${1}" == "description" ]]; then
    echo "Run the DevEnv3 (with preliminary run of the 'down' command)"
    return 0
  fi

  info \
    "The Development Environment will run in a foreground (with a realtime log printing)" \
    "For stop it please press CTRL-C"

  info \
    "For access to your applications please type in the browser:" \
    "http://<application_name>.localhost"

  progress "Starting 'docker-compose down' command"
  run_inside_de3 \
    docker-compose down
  progress "Starting 'docker-compose up' command"
  run_inside_de3 \
    docker-compose up
  progress "Done"
}

main_footer

command_name="${1:-help}"
if ! declare -F "command_${command_name}" >/dev/null; then
  error "Command '${command_name}' is not exists, please run '${DEVENV3_ALIAS} help' or just '${DEVENV3_ALIAS}' command"
fi

shift
command_${command_name} "${@}"
